### PR TITLE
feat: components

### PR DESCRIPTION
**Pull request by enitrat**
A chapter about components

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -1 +1 @@
-scarb 0.7.0
+scarb 2.3.0-rc0
```
Comment (Eikix) [inline .tool-versions:1]: would recommend rc1, it has fixed slow compilation

File: src/ch99-01-05-00-components.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+## Migrating a Contract to a Component
+
+Since both contracts and components share a lot of similarities, it's actually
+very easy to migrate from a contract to a component. The only changes required
+are:
+
+- Adding the `#[starknet::component]` attribute to the module.
+- Adding the `#[embeddable_as(name)]` attribute to the `impl` block that will be
+  embedded in another contract.
+- Adding generic paramters to the `impl` block:
```
Comment (Eikix) [inline src/ch99-01-05-00-components.md:135]:
```suggestion
- Adding generic parameters to the `impl` block:
```

File: src/ch99-01-05-00-components.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  - Adding `TContractState` as a generic parameter.
+  - Adding `+HasComponent<TContractState>` as an impl restriction.
+- Changing the type of the `self` argument in the functions inside the `impl`
```
Comment (Eikix) [inline src/ch99-01-05-00-components.md:138]: would recommend a small minimal example of migration as code snippet after this detailed explanation

File: listings/ch99-starknet-smart-contracts/components/no_listing_01_embeddable/src/lib.cairo [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#[starknet::contract]
+mod simple_contract {
+    #[storage]
+    struct Storage {}
+
+    #[abi(embed_v0)]
+    impl MySimpleImpl = super::SimpleImpl<ContractState>;
+}
+
+use nullable::NullableTrait;
```
Comment (Eikix) [inline listings/ch99-starknet-smart-contracts/components/no_listing_01_embeddable/src/lib.cairo:22]: Why this import, here?

File: src/ch99-01-05-00-components.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+functionality, without having to reimplement the same logic over and over again.
+
+Think of components as Lego blocks. They allow you to enrich your contracts by
+plugging in a module that you or someone else wrote. This module can be a simple
+one, like an ownership component, or more complex like a full-fledged ERC20
+token.
+
+A component is a separate module that can contain storage, events, and
+functions. Unlike a contract, a component cannot be declared or deployed. Its
+logic will eventually be part of the contract’s bytecode it has been embedded
```
Comment (Eikix) [inline src/ch99-01-05-00-components.md:21]: could add that the component's storage will be embedded into the "parent" contract

btw what is the vocab to be adopted?

component and embeddor contract?

component and host contract?

Comment (Eikix) [reply src/ch99-01-05-00-components.md:21]: component and contract?

File: listings/ch99-starknet-smart-contracts/components/no_listing_01_embeddable/src/lib.cairo (hunk 4 above)
Comment (enitrat) [reply listings/ch99-starknet-smart-contracts/components/no_listing_01_embeddable/src/lib.cairo:22]: lol, no idea

File: src/ch99-01-05-00-components.md (hunk 5 above)
Comment (enitrat) [reply src/ch99-01-05-00-components.md:21]: component and embedding contract
contract and embedded component?

**Pull request merged**